api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java

```diff
@@ -0,0 +1,210 @@
+			encounterQuery.setParameter("encounterType", EncounterTypes.TB_SUMMARY_ENCOUNTER.uuid());
+			encounterQuery.setParameter("conceptUuid", tbIdentifierConceptUuid);
+			
+			List<Encounter> patientEncounters = encounterQuery.list();
+			
+			for (Encounter encounter : patientEncounters) {
+				RuleResult<Patient> ruleResult = new RuleResult<>();
+				ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + encounter.getId());
+				ruleResult.setNotes("Client# " + getTbNumber(patient, encounter) + ", has missing " + identifierTitle + " identifier");
+				ruleResult.setEntity(patient);
+				
+				ruleResults.add(ruleResult);
+			}				
+		}
+		
+		return ruleResults;		
+	}	
+		
+	/**
+	 * Patients with no final outcome after 9 months at start of treatment 
+	 * @return List<RuleResult<Patient>>
+	 */
+	public List<RuleResult<Patient>> patientsWithNoFinalOutcomeNineMonthsAfterStartOfTreatment(DateTime date) {
+		log.info("Executing rule to find patients with no final TB Outcome 9 months after start of treatment");
+		
+		String queryString = "SELECT pp.patient FROM PatientProgram pp WHERE pp.voided = 0 AND pp.dateEnrolled <= :date AND pp.outcome IS NULL AND pp.program.uuid = '" + Programs.TB_PROGRAM.uuid() + "'";
+		Query query = getSession().createQuery(queryString);
+        DateTime nineMonthsBefore = date.plusMonths(-9);
+        query.setParameter("date", nineMonthsBefore.toDate());
+	
+		List<Patient> patientsInProgramList = query.list();
+				
+		Set<Patient> uniquePatientList = new HashSet<Patient>(patientsInProgramList);
+		
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		for (Patient patient : uniquePatientList) {
+			//Fetch the latest encounter for this patient
+			String encounterQueryString = "FROM Encounter e WHERE e.patient.id = :patientId AND e.voided = 0 AND e.encounterType.uuid = :encounterTypeUuid ORDER BY e.encounterDatetime DESC";
+			Query encounterQuery = getSession().createQuery(encounterQueryString);
+			encounterQuery.setMaxResults(1);
+			encounterQuery.setParameter("patientId", patient.getId());
+			encounterQuery.setParameter("encounterTypeUuid", EncounterTypes.TB_SUMMARY_ENCOUNTER.uuid());
+			
+			List<Encounter> encounterList = encounterQuery.list();
+				RuleResult<Patient> ruleResult = new RuleResult<>();
+				if (encounterList.size() > 0) {
+					Encounter encounter = encounterList.get(0);
+					ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + encounter);
+					ruleResult.setNotes("Patient #" + getTbNumber(patient, encounter) + " has no final TB Outcome 9 months after start of treatment");
+				} else {
+					ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getId());
+					ruleResult.setNotes("Patient #" + patient.getId() + " has no final TB Outcome 9 months after start of treatment");
```

Comment:
Do not use the System generated ID - at least using the OpenMRS ID since that can be searched for in the Find Patient widget